In [1]:
# Tratamiento de datos
# -----------------------------------------------------------------------
import pandas as pd
import numpy as np

# Visualizaciones
# -----------------------------------------------------------------------
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn import tree



# Para realizar la regresión lineal y la evaluación del modelo
# -----------------------------------------------------------------------
from sklearn.tree import DecisionTreeRegressor, plot_tree
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split,GridSearchCV
from sklearn.metrics import r2_score, mean_absolute_error
import pickle
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error
from category_encoders import TargetEncoder
from sklearn.ensemble import GradientBoostingRegressor



from sklearn.model_selection import KFold,LeaveOneOut, cross_val_score


from sklearn.preprocessing import StandardScaler

from tqdm import tqdm


# Ignorar los warnings
# -----------------------------------------------------------------------
import warnings
warnings.filterwarnings('ignore')
import os 
import sys
sys.path.append("../../")
from src import soporte_ajuste as sa

pd.options.display.float_format = "{:,.2f}".format 
pd.set_option("display.max_columns", None)

In [2]:
df=pd.read_csv("../../datos/toledo_preprocesado.csv", index_col=0)
df

price  propertyType  operation   size  exterior  rooms  bathrooms  \
0  550.00        568.30     577.80  87.00     False      2          2   
1  450.00        568.30     577.80  89.00     False      2          1   
2  450.00        568.30     577.80  60.00      True      2          1   
3  550.00        568.30     577.80  90.00      True      3          1   
4  550.00        568.30     577.80  87.00      True      2          2   
5  750.00        568.30     577.80  95.00      True      3          2   
6  730.00        568.30     577.80  98.00      True      3          2   
7  530.00        568.30     577.80  50.00      True      2          1   
8  650.00        568.30     577.80  70.00      True      2          1   
9  650.00        568.30     577.80  70.00      True      2          1   
10 675.00        568.30     577.80  84.00      True      2          2   
11 650.00        568.30     577.80  84.00      True      2          1   
12 500.00        570.93     577.80  96.00      True      3          2   
13 600.00        568.30     577.80  85.00      True      2          1   
14 750.00        568.30     577.80 149.00      True      2          2   
15 625.00        568.30     577.80  89.00      True      2          2   
16 550.00        568.30     577.80 100.00      True      3          2   
17 500.00        568.30     577.80  90.00      True      3          2   
18 550.00        568.30     577.80 106.00      True      2          1   
19 650.00        568.30     577.80  81.00      True      2          2   
20 550.00        568.30     577.80  90.00      True      3          2   
21 550.00        571.44     577.80  80.00      True      2          1   
22 550.00        568.30     577.80  70.00      True      2          1   
23 650.00        568.30     577.80  87.00      True      2          2   
24 450.00        568.30     577.80  60.00      True      1          1   
25 620.00        568.30     577.80  87.00      True      2          2   
26 400.00        568.30     577.80  55.00      True      1          1   
27 500.00        568.30     577.80 115.00      True      3          2   
28 650.00        568.30     577.80  95.00      True      2          2   
29 550.00        568.30     577.80 105.00      True      3          1   
30 750.00        598.27     577.80 216.00     False      3          2   
31 750.00        568.30     577.80 155.00     False      3          2   
32 700.00        598.27     577.80 238.00     False      4          3   
33 660.00        571.44     577.80  66.00      True      1          1   
34 495.00        568.30     577.80  90.00     False      3          2   
35 470.00        568.30     577.80  80.00     False      1          1   
36 550.00        568.30     577.80  76.00      True      2          1   
39 675.00        568.30     577.80  84.00      True      2          2   
40 340.00        568.30     577.80  50.00      True      1          1   
41 685.00        568.30     577.80  84.00      True      2          2   
42 500.00        570.93     577.80  70.00      True      3          2   
43 460.00        568.30     577.80  85.00      True      1          1   
44 460.00        568.30     577.80  55.00      True      1          1   
45 600.00        570.93     577.80  70.00      True      2          1   
46 400.00        568.30     577.80 100.00      True      3          2   
47 400.00        568.30     577.80 128.00     False      4          2   
49 660.00        568.30     577.80  91.00      True      2          2   
50 500.00        568.30     577.80 100.00      True      3          1   
51 430.00        568.30     577.80  60.00      True      1          1   
52 700.00        598.27     577.80 189.00     False      4          3   
53 700.00        568.30     577.80  97.00      True      2          2   
54 620.00        568.30     577.80  97.00      True      2          2   
55 550.00        568.30     577.80  79.00      True      2          2   
56 650.00        568.30     577.80 176.00      True      3    

In [3]:
x= df.drop(columns="price")
y= df[["price"]]

x_train, x_test, y_train, y_test = train_test_split(x, y, train_size= 0.7, random_state=42)

In [4]:
if isinstance(y, pd.DataFrame):
    y = y.iloc[:, 0]  # Convertir a Serie usando la primera columna

# Dividimos los datos en conjuntos de entrenamiento (train) y prueba (test)
X_train, X_test, y_train, y_test = train_test_split(
    x, y, test_size=0.8, random_state=42
)

# Verificar las formas de los conjuntos de datos resultantes
print(f"Forma de X_train: {X_train.shape}")
print(f"Forma de X_test: {X_test.shape}")
print(f"Forma de y_train: {y_train.shape}")
print(f"Forma de y_test: {y_test.shape}")

# Crear y entrenar el modelo
gbr_model = GradientBoostingRegressor(random_state=42, n_estimators=100, learning_rate=0.1)
gbr_model.fit(X_train, y_train)

# Predicción y evaluación
y_pred_gbr = gbr_model.predict(X_test)

# Calcular el RMSE
rmse_gbr = np.sqrt(mean_squared_error(y_test, y_pred_gbr))

# Calcular el coeficiente R^2
r2_gbr = r2_score(y_test, y_pred_gbr)

# Mostrar resultados
print("Precio real vs Predicción:")
display(pd.DataFrame({'Real': y_test.values, 'Predicción': y_pred_gbr}).head(2))  # Mostrar las primeras 2 filas
print(f"RMSE de Gradient Boosting: {rmse_gbr:.2f}")
print(f"R^2 de Gradient Boosting: {r2_gbr:.2f}")

Forma de X_train: (11, 31)
Forma de X_test: (48, 31)
Forma de y_train: (11,)
Forma de y_test: (48,)
Precio real vs Predicción:


Real  Predicción
0 550.00      545.74
1 750.00      665.36

RMSE de Gradient Boosting: 30.30
R^2 de Gradient Boosting: 0.91


<div style="background-color: #FFBB33; padding: 10px; border-left: 6px solid #000080; color: black; border-radius: 10px;">
RMSE : 30.30

R^2: 0.91
</div>

<div style="background-color: #FBFB; padding: 10px; border-left: 6px solid #000080; color: black; border-radius: 10px;">
Conclusión:  

La tercera métrica (RMSE = 30.30, R² = 0.91) es la mejor opción, ya que combina un bajo error con un alto poder explicativo.
</div>